cmake/OpenCVDetectOpenCL.cmake

```diff
@@ -2,11 +2,30 @@ set(OPENCL_FOUND ON CACHE BOOL "OpenCL library is found")
 if(APPLE)
   set(OPENCL_LIBRARY "-framework OpenCL" CACHE STRING "OpenCL library")
   set(OPENCL_INCLUDE_DIR "" CACHE PATH "OpenCL include directory")
-else(APPLE)
+else()
   set(OPENCL_LIBRARY "" CACHE STRING "OpenCL library")
-  set(OPENCL_INCLUDE_DIR "${OpenCV_SOURCE_DIR}/3rdparty/include/opencl/1.2" CACHE PATH "OpenCL include directory")
+  if(WITH_OPENCL_D3D11_NV)
+    if(NOT EXISTS "${OPENCL_INCLUDE_DIR}/CL/cl_d3d11_ext.h")
+      message(STATUS "CL/cl_d3d11_ext.h not found in OPENCL_INCLUDE_DIR. Automatic detection has been turned on")
```

Comment:
Can we add this header file to `3rdparty/include/opencl`? Is it's license compatible? Is it standardized?

Reply:
Yes, it's compatible and standard. @alalek and I discussed earlier that it relies on `nvopencl64.dll` in the CUDA SDK, so it needs to be detected, and there is no need for HAVE_CUDA.

Reply:
CUDA `.dll` (or `.lib`) is not needed to build OpenCV with support of this extension.

We need just this header file.
But I can't find proper legal source of this file (like [this](https://github.com/KhronosGroup/OpenCL-Headers/tree/master/CL)). Looks like this file comes from CUDA SDK only, under CUDA SDK's own license.

Reply:
Now there is a new problem, we still need to check whether the device has extension function at runtime, otherwise it will throw an uncapable `0xc0000005` exception when executing the `clGetDeviceIDsFromD3D11KHR` function that does not support extension.

> `clGetExtensionFunctionAddressForPlatform` does not guarantee that an extension function is 
actually supported by the platform. The application must also make a corresponding query
 using `clGetPlatformInfo (platform, CL_PLATFORM_EXTENSIONS, ... )` or `clGetDeviceInfo 
(device,CL_DEVICE_EXTENSIONS, ... )` to determine if an extension is supported by the OpenCL 
implementation.
**Notes**
Since there is no way to qualify the query with a device, the function pointer returned must work for all implementations of that extension on different devices for a platform. The behavior of calling a device extension function on a device not supporting that extension is undefined.

[clGetExtensionFunctionAddressForPlatform ](https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clGetExtensionFunctionAddressForPlatform.html)

**A recent commit has fixed this issue.**

---
The license is no different from other header files,and it has been declared at the beginning of the file.
```
* Copyright (c) 2008-2009 The Khronos Group Inc.
* Permission is hereby granted, free of charge...
```
So far, we can choose to add the header file or not, the runtime also needs to detect whether the device supports the corresponding extension.

But I think it's better to keep the status quo, because at compile time, users can clearly know that this extension needs CUDA support.

Reply:
If it has Khronos license lets add it near existed OpenCL header files (3rdparty/include/opencl/1.2/CL).
But keep CMake check for this file, because users can change headers sources via setting of own `OPENCL_INCLUDE_DIR`.

Also line
```
set(OPENCL_INCLUDE_DIR "${OpenCV_SOURCE_DIR}/3rdparty/include/opencl/1.2" CACHE PATH "OpenCL include directory")
```

should be moved before this check (check should be placed after "mark_as_advanced()" statement).

Reply:
Done